functional_tests.go

```diff
@@ -85,14 +87,59 @@ func getFilePath(filename string) (filepath string) {
 	return
 }
 
+// Implements a type of reader which returns a
+// constant stream of random content
+type randomReader struct {
+	TotalBytes       int64
+	RandomSeedChar   string
+	currentReadBytes int
+}
+
+// Implements io.Reader compatible Read() returns random bytes
+// as requested until the upper bound of TotalBytes when
+// Read() will return 0 <= n < len(b) and err is set to io.EOF.
+func (r *randomReader) Read(p []byte) (n int, err error) {
+	if int64(r.currentReadBytes+len(p)) > r.TotalBytes {
+		err = io.EOF
+		// When we have reached the upper bounds we will see
+		// if we need to return n < len(p) this is nil
+		// if all Reads were properly aligned.
+		p = p[:r.TotalBytes-int64(r.currentReadBytes)]
+	}
+
+	// Copy the first random seed char which is
+	// repeated for len(p) bytes.
+	bp := copy(p, []byte(r.RandomSeedChar))
+	for bp < len(p) {
+		copy(p[bp:], p[:bp])
+		bp *= 2
+	}
+
+	// Update how much we returned to the caller.
+	r.currentReadBytes = r.currentReadBytes + len(p)
+	return len(p), err
+}
+
+// Size returns size of the total object.
+func (r *randomReader) Size() int64 {
+	return r.TotalBytes
+}
+
+// Close is just an interface compatible io.Closer but a noop.
+func (r *randomReader) Close() error {
+	return nil
+}
+
 // read data from file if it exists or optionally create a buffer of particular size
-func getDataBuffer(fileName string, size int) (buf []byte) {
+func getDataReader(fileName string, size int) io.ReadCloser {
 	if _, err := os.Stat(getFilePath(fileName)); os.IsNotExist(err) {
-		buf = bytes.Repeat([]byte(string('a')), size)
-		return
+		return &randomReader{
+			TotalBytes:     int64(size),
+			RandomSeedChar: "a",
+		}
```

Comment:
we could remove code for counting bytes read by composing a simple random reader with `io.LimitReader`.
E.g,
```go
  return io.LimitReader{R: randomReader{seed: "a"}, N: size}
```
where

```go
type randomReader struct {
  seed string
}

func (r *randomReader) Read (b []byte) (int, error) {
   return copy(b, bytes.Repeat(r.seed, len(b))), nil
}
```